File: 1. Tic Tac Toe/tic-tac-toe.ts
/**
 * Tic Tac Toe Game Implementation in TypeScript
 * This module implements a simple Tic Tac Toe game with a command-line interface.
 * It includes classes for the game board, players, and the game logic.
 */

enum Symbol {
  X = "X",
  O = "O",
  EMPTY = " ",
}

abstract class Player {
  protected symbol: Symbol;

  constructor(symbol: Symbol) {
    this.symbol = symbol;
  }

  getSymbol(): Symbol {
    return this.symbol;
  }

  abstract makeMove(board: Board): Promise<void>;
}

class HumanPlayer extends Player {
  private readline: any;

  constructor(symbol: Symbol) {
    super(symbol);
    // In a real Node.js environment, you would import readline
    // const readline = require('readline');
    // For this example, we'll simulate user input
  }

  async makeMove(board: Board): Promise<void> {
    while (true) {
      try {
        // In a real implementation, you would use readline to get user input
        // For demonstration, we'll simulate random valid moves
        const availableMoves = this.getAvailableMoves(board);
        if (availableMoves.length === 0) {
          throw new Error("No available moves");
        }

        // Simulate user input (in real implementation, use readline)
        const randomMove =
          availableMoves[Math.floor(Math.random() * availableMoves.length)];
        const row = randomMove.row;
        const col = randomMove.col;

        console.log(`Player ${this.symbol} plays at (${row}, ${col})`);

        if (board.isValidMove(row, col)) {
          board.placeSymbol(row, col, this.symbol);
          break;
        } else {
          console.log("Invalid move. Try again.");
        }
      } catch (error) {
        console.log("Invalid input. Enter valid coordinates.");
      }
    }
  }

  private getAvailableMoves(board: Board): Array<{ row: number; col: number }> {
    const moves: Array<{ row: number; col: number }> = [];
    for (let i = 0; i < board.getSideLength(); i++) {
      for (let j = 0; j < board.getSideLength(); j++) {
        if (board.isValidMove(i, j)) {
          moves.push({ row: i, col: j });
        }
      }
    }
    return moves;
  }
}

class Board {
  private grid: Symbol[][];
  private sideLength: number;

  constructor(sideLength: number = 3) {
    if (sideLength < 3) {
      throw new Error("Board side length must be at least 3");
    }
    this.sideLength = sideLength;
    this.grid = Array(sideLength)
      .fill(null)
      .map(() => Array(sideLength).fill(Symbol.EMPTY));
  }

  getSideLength(): number {
    return this.sideLength;
  }

  display(): void {
    for (let i = 0; i < this.grid.length; i++) {
      console.log(this.grid[i].join("|"));
      if (i < this.sideLength - 1) {
        console.log("-".repeat(2 * this.sideLength - 1));
      }
    }
    console.log(); // Add blank line for better readability
  }

  isValidMove(row: number, col: number): boolean {
    return (
      row >= 0 &&
      row < this.sideLength &&
      col >= 0 &&
      col < this.sideLength &&
      this.grid[row][col] === Symbol.EMPTY
    );
  }

  placeSymbol(row: number, col: number, symbol: Symbol): void {
    if (!this.isValidMove(row, col)) {
      throw new Error(`Invalid move at position (${row}, ${col})`);
    }
    this.grid[row][col] = symbol;
  }

  checkWinner(symbol: Symbol): boolean {
    // Check rows
    for (let i = 0; i < this.sideLength; i++) {
      if (this.grid[i].every((cell) => cell === symbol)) {
        return true;
      }
    }

    // Check columns
    for (let j = 0; j < this.sideLength; j++) {
      if (this.grid.every((row) => row[j] === symbol)) {
        return true;
      }
    }

    // Check main diagonal
    if (this.grid.every((row, i) => row[i] === symbol)) {
      return true;
    }

    // Check anti-diagonal
    if (this.grid.every((row, i) => row[this.sideLength - 1 - i] === symbol)) {
      return true;
    }

    return false;
  }

  isFull(): boolean {
    return this.grid.every((row) => row.every((cell) => cell !== Symbol.EMPTY));
  }

  getWinner(): Symbol | null {
    for (const symbol of [Symbol.X, Symbol.O]) {
      if (this.checkWinner(symbol)) {
        return symbol;
      }
    }
    return null;
  }
}

class Game {
  private board: Board;
  private players: Player[];
  private currentPlayerIdx: number;

  constructor(player1: Player, player2: Player, boardSize: number = 3) {
    if (player1.getSymbol() === player2.getSymbol()) {
      throw new Error("Players must have different symbols");
    }
    this.board = new Board(boardSize);
    this.players = [player1, player2];
    this.currentPlayerIdx = 0;
  }

  private getCurrentPlayer(): Player {
    return this.players[this.currentPlayerIdx];
  }

  private switchPlayer(): void {
    this.currentPlayerIdx = 1 - this.currentPlayerIdx;
  }

  async play(): Promise<void> {
    console.log(
      `Welcome to Tic Tac Toe! Board size: ${this.board.getSideLength()}x${this.board.getSideLength()}`
    );
    console.log("Players take turns. Coordinates are 0-indexed.\n");

    while (true) {
      this.board.display();
      const currentPlayer = this.getCurrentPlayer();

      try {
        await currentPlayer.makeMove(this.board);
      } catch (error) {
        console.log(`Error: ${error}`);
        continue;
      }

      const winner = this.board.getWinner();
      if (winner) {
        this.board.display();
        console.log(`🎉 Player ${winner} wins!`);
        break;
      }

      if (this.board.isFull()) {
        this.board.display();
        console.log("🤝 It's a draw!");
        break;
      }

      this.switchPlayer();
    }
  }
}

// Factory pattern for player creation
class PlayerFactory {
  static createHumanPlayer(symbol: Symbol): HumanPlayer {
    return new HumanPlayer(symbol);
  }
}

// Main execution
async function main(): Promise<void> {
  try {
    const player1 = PlayerFactory.createHumanPlayer(Symbol.X);
    const player2 = PlayerFactory.createHumanPlayer(Symbol.O);
    const game = new Game(player1, player2);
    await game.play();
  } catch (error) {
    console.log(`Error initializing game: ${error}`);
  }
}

// Export for module usage
export { Symbol, Player, HumanPlayer, Board, Game, PlayerFactory };

// Run if this is the main module
// if (require.main === module) {
//   main().catch(console.error);
// }

File: 1. Tic Tac Toe/tic-tac-toe.py
from abc import ABC, abstractmethod
from enum import Enum
from typing import List, Optional

"""Tic Tac Toe Game Implementation
This module implements a simple Tic Tac Toe game with a command-line interface.
It includes classes for the game board, players, and the game logic.
"""

class Symbol(Enum):
    X = 'X'
    O = 'O'
    EMPTY = ' '
    
    def __str__(self):
        return self.value

class Player(ABC):
    def __init__(self, symbol: Symbol):
        self.symbol = symbol

    @abstractmethod
    def make_move(self, board: 'Board') -> None:
        pass

class HumanPlayer(Player):
    def make_move(self, board: 'Board') -> None:
        while True:
            try:
                row = int(input(f"Player {self.symbol}, enter row (0-{board.side_length - 1}): "))
                col = int(input(f"Player {self.symbol}, enter col (0-{board.side_length - 1}): "))
                if board.is_valid_move(row, col):
                    board.place_symbol(row, col, self.symbol)
                    break
                else:
                    print("Invalid move. Try again.")
            except ValueError:
                print(f"Invalid input. Enter numbers 0-{board.side_length - 1}.")

class Board:
    def __init__(self, side_length: int = 3):
        if side_length < 3:
            raise ValueError("Board side length must be at least 3")
        self.side_length = side_length
        self.grid = [[Symbol.EMPTY for _ in range(side_length)] for _ in range(side_length)]

    def display(self) -> None:
        for i, row in enumerate(self.grid):
            print('|'.join(str(cell) for cell in row))
            if i < self.side_length - 1:  # Don't print separator after last row
                print('-' * (2 * self.side_length - 1))
        print()  # Add blank line for better readability

    def is_valid_move(self, row: int, col: int) -> bool:
        return (0 <= row < self.side_length and 
                0 <= col < self.side_length and 
                self.grid[row][col] == Symbol.EMPTY)

    def place_symbol(self, row: int, col: int, symbol: Symbol) -> None:
        if not self.is_valid_move(row, col):
            raise ValueError(f"Invalid move at position ({row}, {col})")
        self.grid[row][col] = symbol

    def check_winner(self, symbol: Symbol) -> bool:
        # Check rows
        for i in range(self.side_length):
            if all(self.grid[i][j] == symbol for j in range(self.side_length)):
                return True
        
        # Check columns
        for j in range(self.side_length):
            if all(self.grid[i][j] == symbol for i in range(self.side_length)):
                return True
        
        # Check main diagonal
        if all(self.grid[i][i] == symbol for i in range(self.side_length)):
            return True
        
        # Check anti-diagonal
        if all(self.grid[i][self.side_length - 1 - i] == symbol for i in range(self.side_length)):
            return True
        
        return False

    def is_full(self) -> bool:
        return all(self.grid[i][j] != Symbol.EMPTY 
                  for i in range(self.side_length) 
                  for j in range(self.side_length))

    def get_winner(self) -> Optional[Symbol]:
        """Check if there's a winner and return the winning symbol."""
        for symbol in [Symbol.X, Symbol.O]:
            if self.check_winner(symbol):
                return symbol
        return None

class Game:
    def __init__(self, player1: Player, player2: Player, board_size: int = 3):
        if player1.symbol == player2.symbol:
            raise ValueError("Players must have different symbols")
        self.board = Board(board_size)
        self.players = [player1, player2]
        self.current_player_idx = 0

    def get_current_player(self) -> Player:
        return self.players[self.current_player_idx]

    def switch_player(self) -> None:
        self.current_player_idx = 1 - self.current_player_idx

    def play(self) -> None:
        print(f"Welcome to Tic Tac Toe! Board size: {self.board.side_length}x{self.board.side_length}")
        print("Players take turns. Enter row and column coordinates (0-indexed).\n")
        
        while True:
            self.board.display()
            current_player = self.get_current_player()
            
            try:
                current_player.make_move(self.board)
            except ValueError as e:
                print(f"Error: {e}")
                continue
            
            winner = self.board.get_winner()
            if winner:
                self.board.display()
                print(f"🎉 Player {winner} wins!")
                break
                
            if self.board.is_full():
                self.board.display()
                print("🤝 It's a draw!")
                break
                
            self.switch_player()

# Factory pattern for player creation
class PlayerFactory:
    @staticmethod
    def create_human_player(symbol: Symbol) -> HumanPlayer:
        return HumanPlayer(symbol)

if __name__ == "__main__":
    try:
        player1 = PlayerFactory.create_human_player(Symbol.X)
        player2 = PlayerFactory.create_human_player(Symbol.O)
        game = Game(player1, player2)
        game.play()
    except ValueError as e:
        print(f"Error initializing game: {e}")
    except KeyboardInterrupt:
        print("\nGame interrupted. Thanks for playing!")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")